pyro/infer/elbo.py

```diff
@@ -62,14 +68,47 @@ def __init__(self,
         self.max_plate_nesting = max_plate_nesting
         self.vectorize_particles = vectorize_particles
         self.retain_graph = retain_graph
-        if self.vectorize_particles:
-            if self.num_particles > 1:
-                if self.max_plate_nesting == float('inf'):
-                    raise ValueError("Automatic vectorization over num_particles requires " +
-                                     "a finite value for `max_plate_nesting` arg.")
-                self.max_plate_nesting += 1
+        if self.vectorize_particles and self.num_particles > 1:
+            self.max_plate_nesting += 1
         self.strict_enumeration_warning = strict_enumeration_warning
 
+    def _guess_max_plate_nesting(self, model, guide, *args, **kwargs):
```

Comment:
It will be nice to be able to use this from HMC too. Is it possible to decouple this from the ELBO class and have guide be optional?

Reply:
I'd recommend just forking and simplifying this method in `HMC`. Two similar methods seem simpler than a complex conglomerate method filled with conditionals, imo.